doc/leas.1
.\" This is the manpage for Leas.
.\" Copyright (c) 2019, Zach Flynn.
.\" See the file COPYING_DOC for copying conditions.
.TH Leas 1 2019-04-26 
.SH NAME
Leas \- the little extensible accounting system \- for keeping accounts in order and studying past spending habits
.SH SYNOPSIS
.B leas [-f FILE] [-l FILE] [-s] [-n] [-v] [-h] [COMMAND]

.SH DESCRIPTION
.B Leas
keeps track of personal accounts and spending.  

.LP
Text entry at the
.B Leas
console follows standard line editor conventions (it has Emacs-like
key bindings, uses the GNU readline library). To cancel a
command, type: Control-C.


.B Leas's
main interactive commands are:

.TP
.BR spend
spend money (transfer from asset to expense account)
.TP
.BR charge
charge a purchase (transfer from liability to expense account)
.TP
.BR earn
earn money (transfer from income to asset account)
.TP
.BR borrow
borrow money (transfer from liability to asset account)
.TP
.BR aa
add account
.TP
.BR at
add transaction to current account
.TP
.BR et\~ [n]
edit transaction, optionally setting how many of the latest transaction numbers to list (any transaction can be chosen whether it is listed or not).
.TP
.BR lt
list most recent transactions in current account (change number to list by modifying the variable leas/number-to-quick-list).
.TP
.BR ltn
list most recent transactions in current account, prompt for number to
list
.TP
.BR ea
rename account
.TP
.BR da
delete account
.TP
.BR dt
delete transaction
.TP
.BR la
list accounts and current balances
.TP
.BR lae
list expense accounts and current balances
.TP
.BR lai
list income accounts and current balances
.TP
.BR laa
list asset accounts and current balances
.TP
.BR lal
list liability accounts and current balances
.TP
.BR bt
total accounts by type (income, expense, asset, liability)
.TP
.BR cex
total across expenditure accounts
.TP
.BR cin
total across income accounts
.TP
.BR cas
total across asset accounts
.TP
.BR cli
total across liability accounts
.TP
.BR cwo
assets - liabilities
.TP
.BR cba
current balance
.TP
.BR re
list transactions in current account matching a regular expression
.TP
.BR sa
set current account
.TP
.BR ca
display current account
.TP
.BR w
write accounts to file
.TP
.BR r
read saved accounts from file
.TP
.BR t
transfer from one account to another
.TP
.BR dtr
delete a transfer
.TP
.BR ltbd
list transactions in current account by day, prompting for the first and last day to show
.TP
.BR v
print version
.TP
.BR sd
set "current" day, changing it allows you to see how account values have changed over time 
.TP
.BR cd
print current day out in YYYY-MM-DD format
.TP
.BR baod
prints account balance over a sequence of days
.TP
.BR exod
prints total expenses as of a sequence of days
.TP
.BR inod
prints total income as of a sequence of days
.TP
.BR liod
prints total liabilities as of a sequence of days
.TP
.BR wood
prints net worth as of a sequence of days
.TP
.BR ttbd
prints total of all transactions between a sequence of days
.TP
.BR ttre
prints total of all transactions between a sequence of days matching a regular expression
.TP
.BR pl
pay a loan, allows you to split a payment between a principal account (liability account) and an interest account (an expense account)
.TP
.BR fn
print out current default save file
.TP
.BR csp
Change the stock price of current account
.TP
.BR "cal [opt]"
print out calendar, can pass a string to cal which acts like specifying options for the program cal at the command line
.TP
.BR p " " X
print X, a Scheme expression
.TP
.BR q
quit, unless -s specified, prompt to save before quit
.LP
The command prompt is a full Guile Scheme interpreter except that you do not need to enclose the outer-most expression in parenthesis. Any valid Scheme code can be executed at the prompt.  For example, typing
.B display \(dqhello\(dq
will print the text "hello".
.B Leas
loads the Scheme functions in the /usr/local/share/leas.scm file at startup. These functions implement the commands above.
.LP
Of particular interest for developing your own functions in Scheme is the function
.B leas/call
which can be used to make Scheme functions interactive.
.LP
The saved file is simply a
.B tar
archive of CSV files that record transactions and characteristics of accounts.  These files are human-readable.
.LP
The prompt is whatever is returned by the evaluating the Scheme expression "(leas/prompt)" so redefining this function will customize your prompt.
.SH OPTIONS
.TP
.BR \-\-file, \-f " " FILE
load accounts from FILE
.TP
.BR \-\-load, \-l " " CODE
execute Scheme code in file CODE
.TP
.BR \-\-no-prompt-on-quit, \-s
do not prompt to save on exit, will not save on exit
.TP
.BR \-\-no-rc, \-n
do not load ~/.leasrc.scm
.TP
.BR \-\-help, \-h
print help message giving command line options.
.TP
.BR \-\-version, \-v
print the version of Leas and copyright info
.TP
.BR COMMAND
execute COMMAND as if it was entered at
.B Leas's
prompt and then exit.  This enables you to, for example, type "leas t" to enter a new transaction (supposing your main Leas file is loaded in your ~/.leasrc.scm).
.SH FILES
.B Leas
executes Scheme code in ~/.leasrc.scm when it starts up if the file is present.
.SH EXAMPLES
.TP
leas
start the Leas interpreter and execute any Scheme code in ~/.leasrc.scm (potentially including code to load data from a default file).
.TP
leas -f myaccounts.leas
load the file 
.B myaccounts.leas
from the current directory.
.TP
leas -s v
quits and exists after displaying version number (without prompting to save data).
.TP
leas aa
displays prompts to add new account after evaluating ~/.leasrc.scm and then exits (prompting to save)
.TP
leas t
displays prompts to add new transaction after evaluating ~/.leasrc.scm and then exits (prompting to save)
.TP
leas -s la
displays current account balances and quits without prompting to save.
.TP
leas -f myaccounts.leas -l myscript.scm -s
Runs a script, presumably to modify myaccounts.leas, and exists.  Leas
can be run noninteractively in this way.
.SH "SEE ALSO"
.SH BUGS
.TP
Please report as an issue to https://github.com/flynnzac/leas
.SH AUTHOR
Zach Flynn <zlflynn@gmail.com>

This manpage is distributed under the GNU Free Documentation License version 1.3.  See COPYING_DOC for the text of the license.  There are no invariant sections.


NEWS
* Initial public version (2020-01-14)

The Little Extensible Accounting System (Leas) is a small program I wrote to manage my finances.  I have used it for over a year and a half daily, and I like it better than other things I have used.  So I thought I would see if others wanted to use it.

I recently did some work to smooth over the rough edges that you are allowed to have in a "personal program".  I have used it for a long enough period of time that I am fairly certain it "works", but I have also been the only user so I am sure there are issues.  Please report them by adding an issue on Github.

src/leas.c
/* 
   Leas - the Little Extensible Accounting System - for keeping accounts in order and studying past spending habits
   Copyright Zach Flynn <zlflynn@gmail.com>

   This program is free software: you can redistribute it and/or
   modify it under the terms of version 3 of the GNU General Public
   License as published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see
   <https://www.gnu.org/licenses/>.
*/


/* The function definitions are only in a header to facilitate testing */
#include "leas.h"

/* Exit Leas */
void
leas_exit ()
{
  char* quit;

  if (leas_prompt_exit)
    {
      quit = readline("Save file? (yes/no) ");
      while ((strcmp(quit,"yes") != 0) &&
             (strcmp(quit, "no") != 0))
        {
          free(quit);
          quit = readline("Save file? (yes/no) ");
        }
	  
      if (strcmp(quit, "yes")==0)
        {
          (void) leas_write(leas_cur_file);
        }
    }
  delete_book(&leas_book);
}

/* Dummy event for signals */
int
dummy_event ()
{
  return 0;
}

/* What to do on interrupt */
void
interrupt_handler (int status)
{
  rl_replace_line("",0);
  if (leas_prompton==PROMPT_SELECT) leas_prompton = PROMPT_COMMAND;
  rl_done = 1;
}

/* Load standard Leas functions */
void
leas_standard_func (char* file)
{
  char* command = copy_string("load ");
  append_to_string(&command, file, "\"");
  (void) exec_string_safe(command);
  free(command);
}


int
main (int argc, char** argv)
{
  SCM ret;
  char* prompt;
  char* command;
  char* fname;
  int i;
  int k;
  time_t curtime_time;
  int load = 1;

  struct option leas_options[] =
    {
     {"version", no_argument, 0, 'v'},
     {"no-rc", no_argument, 0, 'n'},
     {"load", required_argument, 0, 'l'},
     {"file", required_argument, 0, 'f'},
     {"no-prompt-on-quit", no_argument, 0, 's'},
     {"help", no_argument, 0, 'h'}
    };

  int option_index = 0;

  time(&curtime_time);
  leas_curtime = localtime(&curtime_time);

  leas_prompt_exit = 1;
  leas_select_tsct_num = 19;

  scm_with_guile(&register_guile_functions, NULL);
  leas_standard_func(LEAS_SCM_INSTALL);
  leas_cur_file = scm_from_locale_string("_.leas");

  rl_event_hook = dummy_event;
  signal(SIGINT,interrupt_handler);
  
  while ((k = getopt_long(argc, argv, "f:l:snvh", leas_options,
			  &option_index)) != -1)
    {
      switch (k)
        {
        case 'n':
          load = 0;
          break;
        case 'l':
          scm_c_primitive_load(optarg);
          break;
        case 'f':
          leas_cur_file = scm_from_locale_string(optarg);
          fname = scm_to_locale_string(leas_cur_file);

          if (access(fname, R_OK) != -1)
            {
              free(fname);
              fname = scm_to_locale_string(leas_cur_file);
              read_in (fname);
              if (leas_book.n_account > 0)
                leas_cur_acct = scm_from_locale_string
                  (leas_book.accounts[0].name);
            }
          free(fname);
          break;
        case 's':
          leas_prompt_exit = 0;
          break;
        case 'v':
          printf("%s\n", PACKAGE_STRING);
          printf("Copyright (C) 2020 Zach Flynn.\n");
          printf("License GPLv3: GNU GPL version 3 <https://gnu.org/licenses/gpl.html>\n");
          printf("This is free software: you are free to change and redistribute it.\n");
          printf("There is NO WARRANTY, to the extent permitted by law.\n");
          exit(0);
        case 'h':
          printf("Usage: leas [-f FILE] [-l FILE] [-s] [-n] [-v] [-h] [COMMAND]\n");
          printf("Leas %s - the Little Extensible Accounting System\n",
                 PACKAGE_VERSION);
          printf("\n");
          printf("--file,-f FILE  Load Leas save file.\n");
          printf("--load,-l FILE  Load Scheme code at start up.\n");
          printf("--no-prompt-on-quit,-s  Do not prompt to save file on exit.\n");
          printf("--no-rc, -n  Do not load ~/.leasrc.scm.\n");
	  printf("--help,-h Print this help message.\n");
	  printf("--version,-v Print the version of Leas and exit.\n");
          printf("COMMAND Execute Leas command on startup and exit.\n");
	  exit(0);
          break;
        case '?':
          if (optopt=='l')
            fprintf(stderr, "Option -l requires an argument.\n");
          else if (optopt=='f')
            fprintf(stderr, "Option -f requires an argument.\n");
          else
            fprintf(stderr, "Unknown option, -%c.\n", optopt);
	  
          exit(1);
          break;
          break;
        default:
          abort();
          break;
        }
    }

  if (load)
    {
      char* home = getenv("HOME");
      char* leasrc;

      leasrc = copy_string(home);
      append_to_string(&leasrc, "/.leasrc.scm", "");
      
      if (access(leasrc, R_OK) != -1)
        scm_c_primitive_load(leasrc);

      free(leasrc);
    }

  if (optind < argc)
    {
      for (i=optind; i < argc; i++)
        ret = scm_c_catch(SCM_BOOL_T,
                          exec_string_safe_history,
                          argv[i],
                          handle_error,
                          argv[i],
                          NULL,
                          NULL);
      leas_exit();
      return 0;
    }


  if (leas_book.n_account == 0)
    {
      /* create cash account if no other account */
      leas_aa(scm_from_locale_string("Cash"),
              scm_from_locale_string("asset"),
              scm_from_double(0.0));
    }
  
  leas_prompton = PROMPT_COMMAND;

  while (leas_prompton != PROMPT_OFF)
    {
      ret = scm_c_catch(SCM_BOOL_T,
                        exec_string_safe,
                        "leas/prompt",
                        handle_error,
                        "leas/prompt",
                        NULL,
                        NULL);

      prompt = scm_to_locale_string(ret);
      command = readline(prompt);
      
      ret = scm_c_catch(SCM_BOOL_T,
                        exec_string_safe_history,
                        command,
                        handle_error,
                        command,
                        NULL,
                        NULL);
      
      free(prompt); free(command);
    }
  scm_gc();
  leas_exit();
  return 0;
}

